cardano-testnet/src/Testnet/Runtime.hs

```diff
@@ -204,14 +224,30 @@ startLedgerNewEpochStateLogging testnetRuntime tmpWorkspace = withFrozenCallStac
           Api.QuickValidation
           (EpochNo maxBound)
           ()
-          (\epochState _ _ -> handler logFile epochState)
+          (\epochState _ _ ->
+              liftIO $ evalStateT (handler logFile epochState) 0
+          )
       H.note_ $ "Started logging epoch states to: " <> logFile
+
+      -- In order to create a diff of the epoch state logging file contents
+      -- we must do so when the resources are deallocated (see runInBackground)
+      -- and therefore when the log file is no longer in use.
+      void . H.evalM $ allocate (pure ()) $ \_ -> do
```

Comment:
You can also catch `AsyncCancelled` in the `handler` and act upon it instead. That may be easier than installing cleanups in `MonadResource`.

Reply:
So I found out that technically `allocate` is the wrong thing to use here because we are forking a thread. See my most recent changes + comments.